File: 0486-predict-the-winner/0486-predict-the-winner.cpp
class Solution {
public:
    int maxDiff(int left,int right,vector<int> &nums)
    {
        if(left == right)
        {
            return nums[left];
        }
        int scoreL = nums[left]-maxDiff(left+1,right,nums);
        int scoreR = nums[right] - maxDiff(left,right-1,nums);

        return max(scoreL,scoreR);
    }
    bool predictTheWinner(vector<int>& nums) {
        
        int n = nums.size();

        if(n%2 == 0)
        {
            return true;
        }
        return maxDiff(0,n-1,nums) >= 0;

        
    }
};

File: 0486-predict-the-winner/0486-predict-the-winner.py
from functools import cache
class Solution:

    def predictTheWinner(self, nums: List[int]) -> bool:
        i = 0
        j = len(nums)-1
        @cache
        def solve(i,j):
            if i == j:
                return nums[i]
            choice_a = nums[i]  - solve(i+1,j)
            choice_b = nums[j] - solve(i,j-1)
            return max(choice_a,choice_b)
        score_d = solve(i,j)
        return score_d >= 0
        
        